src/stan/lang/generator/get_block_var_dims.hpp

```diff
@@ -0,0 +1,44 @@
+#ifndef STAN_LANG_GENERATOR_GET_BLOCK_VAR_DIMS_HPP
+#define STAN_LANG_GENERATOR_GET_BLOCK_VAR_DIMS_HPP
+
+#include <stan/lang/ast.hpp>
+#include <stan/lang/generator/constants.hpp>
+#include <ostream>
+#include <vector>
+
+namespace stan {
+  namespace lang {
+
+    /**
+     * Return vector of size expressions for all dimensions
+     * of a block_var_decl in the following order:
+     * matrix cols (if matrix type),
+     * matrix row / row_vector / vector length (if matrix/vec type),
+     * array dim N through array dim 1
+     *
+     * @param[in] decl block_var_decl
+     */
+    std::vector<expression>
+    get_block_var_dims(const block_var_decl decl) {
+      std::vector<expression> dims;
+
+      block_var_type bt = decl.type();
+      if (bt.is_array_type())
+        bt = bt.array_contains();
+      if (bt.bare_type().is_matrix_type()) {
```

Comment:
[optional]
A super-helpful method on bare types would be to return all their dimensions:
```
struct block_var_type {
  ...
  std::vector<expression> dimensions() const;
  ...
}
```
Then I think all this could be replaced with
```
std::vector<expression> dims = decl.type().dimensions();
```
I think it's used elsewhere, too.  I'm not sure how this is all going to work out when we have tuples, as they don't have traditional dimensions like this.